qemu/tests/vsock_test.py

```diff
@@ -0,0 +1,149 @@
+import aexpect
+
+from avocado.utils import process
+
+from virttest import data_dir
+from virttest import error_context
+from virttest import utils_misc
+
+
+def compile_nc_vsock(vm, session):
+    """
+    Copy and compile nc-vsock on both host and guest
+
+    :param vm: Object qemu_vm.VM
+    :param session: vm session
+    :return: Path to binary nc-vsock or None if compile failed
+    """
+    nc_vsock_dir = '/home/'
+    nc_vsock_bin = 'nc-vsock'
+    nc_vsock_c = 'nc-vsock.c'
+    src_file = os.path.join(data_dir.get_deps_dir("nc_vsock"), nc_vsock_c)
+    rm_cmd = 'rm -rf %s/%s*' % (nc_vsock_dir, nc_vsock_bin)
+    session.cmd(rm_cmd)
+    process.system(rm_cmd, shell=True, ignore_status=True)
+    cmd_cp = "cp %s %s" % (src_file, nc_vsock_dir)
+    process.system(cmd_cp)
+    vm.copy_files_to(src_file, nc_vsock_dir)
+    compile_cmd = "cd %s && gcc -o %s %s" % (
+        nc_vsock_dir, nc_vsock_bin, nc_vsock_c)
+    if process.system(compile_cmd, shell=True) != 0:
+        return None
+    if session.cmd_status(compile_cmd) != 0:
+        return None
+    return '%s/%s' % (nc_vsock_dir, nc_vsock_bin)
+
+
+def nc_vsock_listen(nc_vsock_bin, port, session):
+    """
+    Open nc-vsock listening process from guest, cmd: nc-vsock -l $port
+
+    :param nc_vsock_bin: path of binary nc-vsock
+    :param port: the port to listen
+    :param session: guest shell session
+    :return: the shell session with opened vsock listening process
+    """
+    nc_vsock_cmd = "%s -l %s" % (nc_vsock_bin, port)
+    session.read_nonblocking(0, timeout=10)
+    logging.info("Listening to the vsock port from guest: %s" % nc_vsock_cmd)
+    session.sendline(nc_vsock_cmd)
+
+
+def nc_vsock_connect(nc_vsock_bin, guest_cid, port):
+    """
+    Connect to vsock port from host, cmd: nc-vsock $guest_cid $port
+
+    :param nc_vsock_bin: path of binary nc-vsock
+    :param guest_cid: guest cid to connect
+    :param port: port to connect
+    :return: The vsock session from host side, being waiting for input
+    """
+    nc_vsock_cmd = "%s %s %s" % (nc_vsock_bin, guest_cid, port)
+    logging.info("Connect to the vsock port on host: %s" % nc_vsock_cmd)
+    return aexpect.Expect(
+        nc_vsock_cmd,
+        auto_close=False,
+        output_func=utils_misc.log_line,
+        output_params=("vsock_%s_%s" % (guest_cid, port),))
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Vsock basic function test
+
+    1. Boot guest with vhost-vsock-pci device
+    2. Download and compile nc-vsock on both guest and host
+    3. Start listening inside guest, nc-vsock -l $port
+    4. Connect guest CID from host, nc-vsock $guest_cid $port
+    5. Input character, e.g. 'Hello world'
+    6. Check if guest receive the content correctly
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment
+    """
+
+    def clean():
+        """ Clean the environment """
```

Comment:
why not put an argument in `clean` to decide if`tmp_file` should be deleted.

Reply:
I think it does not mean much,  As the problem is we can't call a undefined param, nothing to do with the value of the param.   The last version (define the param before use it ) is enough, no need to make it so complex.